jctools-experimental/src/main/java/org/jctools/stacks/WFIStack.java

```diff
@@ -0,0 +1,329 @@
+ * These node objects cannot be reused, and must be allocated fresh for every push.
+ * <p>
+ * The implementation draws inspiration from Dmitry Vyukov's
+ * <a href="http://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue">
+ * Intrusive MPSC node-based queue</a>, but is oriented as a stack instead of a queue, and supports MPMC.
+ *
+ * @param <T> The type of node in this stack.
+ */
+@SuppressWarnings({"unchecked", "NullableProblems"})
+public class WFIStack<T extends WFIStack.Node> implements Iterable<T>
+{
+    /**
+     * The super-class of all nodes, or entries, for {@link WFIStack wait-free intrusive stacks}.
+     */
+    public static abstract class Node
+    {
+        volatile Node next;
+        int count;
+
+        final Node next()
+        {
+            Node n = next;
+            while (n == null)
+            {
+                // TODO: Thread.onSpinWait
+                Thread.yield();
+                n = next;
+            }
+            return n;
+        }
+
+        <X extends Node> X self()
+        {
+            return (X) this;
+        }
+    }
+
+    private static class NodeIterable<E extends Node> implements Iterable<E>
+    {
+        private Node start;
+
+        private NodeIterable(Node start)
+        {
+            this.start = start;
+        }
+
+        @Override
+        public Iterator<E> iterator()
+        {
+            return new NodeIterator<E>(start);
+        }
+
+        public NodeIterable<E> reverse()
+        {
+            Node curr = END;
+            Node next = start;
+            while (next != END)
+            {
+                Node n = next;
+                next = next.next();
+                n.next = curr;
+                curr = n;
+            }
+            start = curr;
+            return this;
+        }
+    }
+
+    private static class NodeIterator<E extends Node> implements Iterator<E>
+    {
+        Node next;
+
+        public NodeIterator(Node start)
+        {
+            next = start.self();
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return next != null;
+        }
+
+        @Override
+        public E next()
+        {
+            Node n = next;
+            if (n == null)
+            {
+                throw new NoSuchElementException();
+            }
+            next = n.next().self();
+            return (E) n;
+        }
+    }
+
+    private static final Node END = new Node()
+    {
+        {
+            next = this; // Next from END is END itself.
+        }
+
+        @Override
+        <X extends Node> X self()
+        {
+            return null;
+        }
+    };
+    private static final long HEAD_OFFSET = UnsafeAccess.fieldOffset(WFIStack.class, "head");
+
+    @SuppressWarnings( {"FieldCanBeLocal", "FieldMayBeFinal"})
+    private volatile Node head; // Accessed via UNSAFE.
```

Comment:
I'm not a big fan of padding, especially manual padding, but probably I would pad this from any heap noisy neighbours, including the obj header itself